File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import sys
from collections import deque
dx = [0,0,1,-1]
dy = [1,-1,0,0]
def bfs(maps) :
    dQ = deque()
    maps[0][0] = 1 #방문 표시
    dQ.append((0, 0))
    xend = len(maps)
    yend = len(maps[0])
    while dQ : 
        x, y = dQ.popleft()
        if x == xend-1 and y == yend-1 :
            return maps[x][y]
        for i in range(4) :
            xx = x + dx[i]
            yy = y + dy[i]
            if 0<=xx<xend and 0<=yy<yend :
                if maps[xx][yy] == 1 :
                    maps[xx][yy] = maps[x][y] + 1
                    dQ.append((xx, yy))
    return -1
            
    
def solution(maps):
    answer = 0
    xend = len(maps)
    yend = len(maps[0])
     
    return bfs(maps)

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    int[] dx = {0, 0, 1, -1};
    int[] dy = {1, -1, 0, 0};
    
    public int solution(int[][] maps) {
        int answer = 0;
        int[][] visited = new int[maps.length][maps[0].length];
        // System.out.println(visited);
        bfs(maps, visited);
        answer = visited[maps.length-1][maps[0].length-1];
        
        if(answer == 0)
            return -1;
        else
            return answer;
    }
    
    public void bfs(int[][] maps, int[][] visited) {
        int x = 0;
        int y = 0;
        visited[x][y] = 1;
        Queue<int[]> queue  = new LinkedList<>();
        queue .add(new int[]{x, y});
        
        while(!queue.isEmpty()) {
            int[] current = queue.remove();
            int cx = current[0];
            int cy = current[1];
            
            for(int i = 0; i < 4; i++) {
                int xx = cx + dx[i];
                int yy = cy + dy[i];
                
                if(xx < 0 || xx >= maps.length || yy < 0 || yy >= maps[0].length)
                    continue;
                if(visited[xx][yy] == 0 && maps[xx][yy] == 1) {
                    visited[xx][yy] = visited[cx][cy] + 1;
                    queue.add(new int[]{xx, yy});
                }
            }
        }
    }
    
}